Warning at multistage build

Gives next warnings dockerfilelint:
    15:1    warning dockerfilelint: Clarity      Base images should specify a tag to use.
    21:1    warning dockerfilelint: Clarity      Base images should specify a tag to use.
    26:1    warning dockerfilelint: Clarity      Base images should specify a tag to use.

Where should I put this tag?

```
# docker --host=:3000 build -t spa .

# ---- Base image ----
FROM node:alpine AS base

LABEL version="1.0"
LABEL description="HRketing SPA application"
LABEL maintainer="Alex Kostyukov <kostyukov@itopcase.ru>"

RUN apk --no-cache --update add git
WORKDIR /usr/src/deps
COPY package*.json ./

# ---- Dependencies ----
FROM base AS dependencies
RUN npm set progress=false && npm config set depth 0
RUN npm install --loglevel=error --only=production --prefix node_modules_production
RUN npm install --loglevel=error

# ---- Linters ----
FROM dependencies AS test
COPY . .
RUN npm run lint

# ---- Release ----
FROM base AS release

ARG NODE_ENV=production
ENV NODE_ENV=$NODE_ENV

WORKDIR /usr/src/app
COPY --from=dependencies /usr/src/deps/node_modules_production ./node_modules
COPY . .
EXPOSE 3000
CMD ["npm", "run", "deploy"]
```

--- a/lib/checks.js
+++ b/lib/checks.js
@@ -29,7 +29,7 @@ var commands = module.exports = {
     return result;
   },
 
-  base_image_tag: function(args) {
+  base_image_tag: function(args, previousStageNames) {
     // scratch is a special base layer for Docker, it means there is no base layer
     if (args === 'scratch') {
       return [];
@@ -42,7 +42,10 @@ var commands = module.exports = {
     }
     var result = [];
     if (baseImage.length === 1) {
-      result.push('missing_tag');
+      // support for multistage builds, a base image can reference a previous stage
+      if (!previousStageNames || previousStageNames.indexOf(baseImage[0].toLowerCase()) === -1) {
+        result.push('missing_tag');
+      }
     } else if (baseImage[baseImage.length - 1] === 'latest') {
       result.push('latest_tag');
     }

--- a/lib/index.js
+++ b/lib/index.js
@@ -52,6 +52,7 @@ module.exports.run = function(configPath, content) {
   var state = {
     stages: [{
       instructionsProcessed: 0,
+      name: "",
       cmdFound: false // this doesn't seem to be used anywhere
     }],
     items: [],
@@ -65,10 +66,19 @@ module.exports.run = function(configPath, content) {
     // We care about only having 1 cmd instruction
     if (result.command === 'cmd') {
       state.stages[state.stages.length - 1].cmdFound = true;
-    } else if (result.command === 'from' && state.stages[state.stages.length - 1].instructionsProcessed !== 0) {
+    } else if (result.command === 'from') {
       // Each FROM command constitutes a new stage in a multistage build
+      const instructionParts = instructions[idx].split(' ');
+      let stageName = '';
+      if (instructionParts.length > 2) {
+        if (instructionParts[instructionParts.length - 2].toLowerCase() === 'as') {
+          stageName = instructionParts[instructionParts.length - 1].toLowerCase();
+        }
+      }
+
       state.stages.push({
         instructionsProcessed: 0,
+        name: stageName,
         cmdFound: false
       });
     }
@@ -146,7 +156,10 @@ function runLine(state, instructions, idx) {
     // Vaildate each command individually
     switch (cmd) {
       case 'from':
-        checks.base_image_tag(args).forEach(function(item) {
+        const previousStageNames = state.stages.map(function(stage) {
+          return stage.name;
+        });
+        checks.base_image_tag(args, previousStageNames).forEach(function(item) {
           items.push(messages.build(state.rules, item, line));
         });
         break;

--- a/test/examples/Dockerfile.from-latest-multistage
+++ b/test/examples/Dockerfile.from-latest-multistage
@@ -1,0 +1,7 @@
+FROM golang:latest as builder
+
+FROM builder
+RUN ls -l /
+
+FROM alpine:latest
+CMD ["./app"]

--- a/test/index.js
+++ b/test/index.js
@@ -40,6 +40,17 @@ describe("index", function(){
     });
   });
 
+  describe("#from-latest-multistage", function(){
+    it("validates the from-latest-multistage Dockerfile has 3 issues", function(){
+      const actual = dockerfilelint.run('./test/examples', fs.readFileSync('./test/examples/Dockerfile.from-latest-multistage', 'UTF-8'));
+      expect(actual).to.have.length(2);
+      expect(actual[0].line).to.equal(1);
+      expect(actual[0].rule).to.equal('latest_tag');
+      expect(actual[1].line).to.equal(6);
+      expect(actual[1].rule).to.equal('latest_tag');
+    });
+  });
+
   describe("#redis", function(){
     it("validates the redis Dockerfile has no issues", function(){
       expect(dockerfilelint.run('./test/examples', fs.readFileSync('./test/examples/Dockerfile.redis', 'UTF-8'))).to.be.empty;

